# Import necessary packages

In [2]:
import tensorflow as tf

2022-05-30 16:44:04.072467: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2022-05-30 16:44:04.075467: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2022-05-30 16:44:04.075475: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.


# import mnist data from keras

In [13]:
mnist_dataset = tf.keras.datasets.mnist
mnist_dataset

<module 'keras.api._v2.keras.datasets.mnist' from '/home/nsl56/.local/lib/python3.8/site-packages/keras/api/_v2/keras/datasets/mnist/__init__.py'>

# load and preprocess data : 

In [14]:
(x_train, y_train) , (x_test, y_test) = mnist_dataset.load_data()

# define your model

In [15]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(512, activation=tf.nn.relu),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])


# **compile the model with specifying optimizer, loos and metrics**

In [16]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# **train the model**

In [17]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 2s 1ms/step - loss: 2.7148 - accuracy: 0.8782
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4471 - accuracy: 0.9042
Epoch 3/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4090 - accuracy: 0.9117
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3884 - accuracy: 0.9181
Epoch 5/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3789 - accuracy: 0.9213


# **evaluate the model on test data**

In [18]:
loss, accuracy = model.evaluate(x_test, y_test)

313/313 [==============================] - 0s 607us/step - loss: 0.2836 - accuracy: 0.9429


In [19]:
print("loss: {} accuracy: {}".format(loss, accuracy))

loss: 0.2835697829723358 accuracy: 0.9429000020027161


Our model got 94.2 % accuracy, which is good. But let's observe what happens if we normalize the train and test input

# normalize train and test input data :

For faster convergence, we often normalize the data. 
Since the images are 28 X 28 pixels, the values are in the 
range [0,255] . So, for this dataset, we normalize the input data (both train and test) by 255

In [20]:
def normalize(data, value) :
    return data / value

In [21]:
value = 255.0
x_train = normalize(x_train , value)
x_test  = normalize(x_test , value)

# Let's define our model again with the modified values of train and test input

In [33]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(512, activation=tf.nn.relu),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])


# **compile the model with specifying optimizer, loos and metrics**

In [34]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# **train the model**

In [35]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2204 - accuracy: 0.9333
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0971 - accuracy: 0.9693
Epoch 3/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0703 - accuracy: 0.9779
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0532 - accuracy: 0.9833
Epoch 5/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0443 - accuracy: 0.9861


# **evaluate the model on test data**

In [36]:
loss, accuracy = model.evaluate(x_test, y_test)

313/313 [==============================] - 0s 582us/step - loss: 0.0707 - accuracy: 0.9786


In [37]:
print("loss: {} accuracy: {}".format(loss, accuracy))

loss: 0.07073397189378738 accuracy: 0.978600025177002


Before normalization, we got accuracy of 94.2 % , But After normalizing the training and test inputs ,
we can clearly see that accuracy improved (close to 97.8% !!) 

Let's experiment more with our model. Let's see what happens if we increase the number of epoches .

In [38]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(512, activation=tf.nn.relu),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])


In [39]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [40]:
model.fit(x_train, y_train, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2191 - accuracy: 0.9353
Epoch 2/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0977 - accuracy: 0.9700
Epoch 3/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0694 - accuracy: 0.9781
Epoch 4/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0535 - accuracy: 0.9831
Epoch 5/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0433 - accuracy: 0.9858
Epoch 6/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0360 - accuracy: 0.9883
Epoch 7/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0310 - accuracy: 0.9896
Epoch 8/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0269 - accuracy: 0.9906
Epoch 9/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0251 - accuracy: 0.9915
Epoch 10/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.022

In [41]:
loss, accuracy = model.evaluate(x_test, y_test)

313/313 [==============================] - 0s 581us/step - loss: 0.0793 - accuracy: 0.9804


In [42]:
print("loss: {} accuracy: {}".format(loss, accuracy))

loss: 0.07930212467908859 accuracy: 0.980400025844574


We can see that, model's performance improved from 97.8 % to 98 % 

In [43]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(512, activation=tf.nn.relu),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])


In [44]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [45]:
model.fit(x_train, y_train, epochs=15)

Epoch 1/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2193 - accuracy: 0.9349
Epoch 2/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0975 - accuracy: 0.9696
Epoch 3/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0690 - accuracy: 0.9779
Epoch 4/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0532 - accuracy: 0.9827
Epoch 5/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0435 - accuracy: 0.9859
Epoch 6/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0372 - accuracy: 0.9876
Epoch 7/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0308 - accuracy: 0.9900
Epoch 8/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0274 - accuracy: 0.9907
Epoch 9/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0239 - accuracy: 0.9922
Epoch 10/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.022

In [46]:
loss, accuracy = model.evaluate(x_test, y_test)

313/313 [==============================] - 0s 583us/step - loss: 0.0770 - accuracy: 0.9826


In [47]:
print("loss: {} accuracy: {}".format(loss, accuracy))

loss: 0.0769941583275795 accuracy: 0.9825999736785889


Again we can see that , increasing epoches from 10 to 15, models accuracy increases from 98% to 98.26% !

# Reference :

https://www.youtube.com/watch?v=6_upGYUtjdU

https://www.quora.com/Why-are-image-pixel-values-zero-centered-by-dividing-by-255-before-passing-them-as-input-to-feedforward-neural-network-or-convolutional-NN

https://medium.com/analytics-vidhya/a-tip-a-day-python-tip-8-why-should-we-normalize-image-pixel-values-or-divide-by-255-4608ac5cd26a